lib/pbench/server/api/resources/datasets_list.py

```diff
@@ -294,61 +294,65 @@ def filter_query(filters: list[str], query: Query) -> Query:
 
             {
-                "dataset": {"name": {}, "metalog": {"pbench": {"script": {}}}},
-                "server": {"deletion": {}, "tarball-path": {}},
-                "global": {"server": {"legacy": {"sha1": {}}}}
+                "dataset": {"name": None, "metalog": {"pbench": {"script": None}}},
+                "server": {"deletion": None, "tarball-path": None},
+                "global": {"server": {"legacy": {"sha1": None}}}
             }
 
         Args:
-            keys: a JSONOBJECT to update with the recursive key/value
+            aggregate: a JSONOBJECT to update with the recursive key/value
             key: the current metadata key path element
             value: the current metadata key's value
         """
-        if key in keys:
-            p = keys[key]
-        else:
-            p = {}
-            keys[key] = p
         if isinstance(value, dict):
+            p = aggregate.get(key)
+            if p is None:
+                p = {}
+                aggregate[key] = p
             for k, v in value.items():
                 self.accumulate(p, k, v)
+        elif key not in aggregate:
+            aggregate[key] = None
 
     def keyspace(self, query: Query) -> JSONOBJECT:
         """Aggregate the dataset metadata keyspace
 
         Run the query we've compiled, but instead of returning Dataset proxies,
         we only want the metadata key/value pairs we've selected.
 
-        NOTE: In general we expect every dataset to have some metadata; however
-        in the event one doesn't, the key and value returned by the joined
-        query will be None. We handle this by ignoring those values.
+        NOTE: The SQL left outer join returns a row for each row in the "left"
+        table (Dataset) even if there is no matching foreign key in the "right"
+        table (Metadata). This means a dataset with no metadata will result in
+        a join row here with key and value of None. The `elif` in the loop will
+        silently ignore rows with a null key to handle this case.
 
         Args:
             query: The basic filtered SQLAlchemy query object
 
         Returns:
             The aggregated keyspace JSON object
         """
-        keys: JSONOBJECT = {"dataset": {c.name: {} for c in Dataset.__table__._columns}}
+        aggregate: JSONOBJECT = {
+            "dataset": {c.name: None for c in Dataset.__table__._columns}
+        }
         list = query.with_entities(Metadata.key, Metadata.value).all()
         for k, v in list:
             # "metalog" is a top-level key in the Metadata schema, but we
             # report it as a sub-key of "dataset".
             if k == Metadata.METALOG:
-                self.accumulate(keys["dataset"], k, v)
+                self.accumulate(aggregate["dataset"], k, v)
             elif k:
-                self.accumulate(keys, k, v)
-        return keys
+                self.accumulate(aggregate, k, v)
+        return aggregate
```

Comment:
I'm still having trouble with my understanding.  🙁  The last sentence in the NOTE says that the `elif` will ignore rows with "a null key" (which seems to match the code), but the preceding sentence says that it is the _value_ which might be `None`...which one is wrong, the comment, the code, or my understanding?  😆

Reply:
The docstring says "key and value". The SQL left outer join will fill in all columns selected for a missing row in the right column with null: so since I'm asking for key and value, they'll both be null. However, a null `value` for a non-null `key` is fine (it could be set that way deliberately, and there's nothing in the APIs or database to filter that out), so I don't worry about that: it's got nothing to do with the aggregation.

Apparently my explanation still isn't good enough to avoid confusion, so I'll take another look at it. 😦

Reply:
I'm sorry that my grasp on things here is so poor.  `list` is a list of key-value pairs, one pair for each `Dataset` in the query, right?  And, when we unpack it into `k` and `v`, what exactly is `k` if we only have one pair for each dataset??  Based on lines 344 & 351, it seems like `k` is a column in one of the two joined tables...so I seem to have an off-by-one problem in the hierarchy.  😞

Reply:
Hmm. OK, let me break this down as much as I can; maybe too much detail, but perhaps even so that'll help someone.

A SQL Join creates a new "scratch" table combining columns from two (or more) permanent tables. In this case, we're joining Dataset and Metadata.

We start out with `query = Database.db_session.query(Dataset).outerjoin(Metadata)`; that sets up a select statement that builds a table combining Dataset (left) and Metadata (right) into a theoretical table with all of the columns from each, connected by the Metadata foreign key (`dataset_ref`) matching the Dataset `id`.

So each row of the theoretical join table has `id`, `owner_id`, `name`, `uploaded`, `access` from the (left) dataset table and `id`, `dataset_ref`, `user`, `key`, and `value` from the (right) metadata table.

In a default inner join, the join table has a row only where there is a `Metadata.dataset_ref == Dataset.id`. "Orphaned" rows (without a partner) on *either side* are omitted. That means the join table will have no row for a dataset row with no matching metadata, *or* for a metadata row without matching dataset. That's why I had to switch to a left outer join.

In a left outer join, we get all of the matching rows, but we *also* get unmatched rows from the *left* table (dataset) even if there's no matched row on the right, but SQL will fill those columns in the join table with null. In a right outer join, it'll do this for a missing row on either side. (Since we cascade delete metadata rows with the matching dataset through the foreign key, if we ever do get orphaned metadata rows they're an anomaly, they're not interesting to us, and if we ever wanted to make an attempt to clean them up it wouldn't be in the main API code.)

Now, the interesting thing about SQL is that a row can't have multiple values for a column. That's why we have foreign keys and linked tables. In the join table, if we have two metadata keys linked to the same dataset, that dataset appears in *two* rows of the joined table: the `Dataset` column values are all duplicated, but the `Metadata` columns in each row will be unique.

Now we get to SQLAlchemy. The `query(Dataset)` is part of the SQLAlchemy ORM that implicitly selects *all* the `Dataset` columns as if we'd done `SELECT id, name, owner_id, resource_id, access, uploaded`. It implicitly dedups on `.all()` and constructs ORM proxy objects. Because we're doing the `.outerjoin(Metadata)`, in addition to the default `LEFT OUTER JOIN ON`, it adds the Metadata `id`, `user_id`, `key`, and `value` columns to the `SELECT`. Duplicated Dataset columns in a row mean we matched more than one Metadata, so *those* columns get collected into `Metadata` proxy objects and linked into the `metadata` collection on the deduped Dataset.

For the aggregator, I don't really care about the datasets, and I have no interest in either `Dataset` or `Metadata` proxy objects; but I want all of the metadata linked to any dataset that the `/datasets` filters would have returned. So here's the "tricky" part. The `keyspace` method *overrides* SQLAlchemy's default column selector with `.with_entities(Metadata.key, Metadata.value)`. That allows me to use the same query builder so I don't have to worry about duplicating the filters, while minimizing and customizing the data SQL has to return from that query.

"Logically" SQL still builds the same join table, however I'm only populating those two columns from the `Metadata` (right) table. If a dataset has four `Metadata` rows (`metalog`, `server`, `global`, and `user`), this will result in *four* key/value pairs in the sequence returned by `.all()`. We don't actually know to which dataset each belongs, because in this context we don't care, and I keep the SQL as light as possible.

So, the `.all()` sequence gives us a k/v pair for each metadata row that's linked to any dataset that would have been returned by the set of `/datasets` filters without the `keysummary` parameter. There'll be 0 to 4 rows for each dataset, depending on which namespaces are populated: and we're just aggregating the key namespace across all of those rows.

Reply:
Thanks, Dave!  That was awesome!!


> if there's no matched row on the right, but SQL will fill those columns in the join table with null

Check.

> a row can't have multiple values for a column. [...] In the join table, if we have two metadata keys linked to the same dataset, that dataset appears in two rows of the joined table

Check.

> Duplicated Dataset columns in a row mean we matched more than one Metadata, so those columns get collected into `Metadata` proxy objects and linked into the `metadata` collection on the deduped Dataset.

Not sure what this means, but I don't _think_ it's important....

> I'm only populating those two columns from the `Metadata` (right) table.

I suspected as much....

> If a dataset has four `Metadata` rows (`metalog`, `server`, `global`, and `user`), this will result in _four_ key/value pairs in the sequence returned by `.all()`. We don't actually know to which dataset each belongs

Ah ha!  🌤️

> There'll be 0 to 4 rows for each dataset, depending on which namespaces are populated

Uh-oh...  ☁️  How do we end up with null keys, again?  We do a left join of a dataset which has no metadata, we then select the `Metadata.key` and `Metadata.value` columns which have been null-filled, and that yields a null key?...but, then why do we get rows with null keys instead of zero rows?  When would we get less than four rows for a given dataset?

Thanks!

Reply:
> > Duplicated Dataset columns in a row mean we matched more than one Metadata, so those columns get collected into `Metadata` proxy objects and linked into the `metadata` collection on the deduped Dataset.
> 
> Not sure what this means, but I don't _think_ it's important....

If there are two rows in the join table with the same Dataset columns (particularly `Dataset.id`), then we're representing two `Metadata` rows connected to the same `Dataset`. So we get a single `Dataset` proxy object with two `Metadata` proxy objects on its `Dataset.metadata` "relationship" attribute list.

> > There'll be 0 to 4 rows for each dataset, depending on which namespaces are populated
> 
> Uh-oh... cloud How do we end up with null keys, again? We do a left join of a dataset which has no metadata, we then select the `Metadata.key` and `Metadata.value` columns which have been null-filled, and that yields a null key?...

Right.

> but, then why do we get rows with null keys instead of zero rows? When would we get less than four rows for a given dataset?

In terms of the `SELECT`, I think the `JOIN` has already been done, and it's constructed a logical join table with all the rows and columns: I don't think the fact that we're selecting only two columns, which happen to have been fabricated nulls, makes any difference in that process. (It shouldn't, really, because in other contexts the distinction could be important.)

With the left outer join we *always* get a row for each `Dataset` even if it has no metadata. That's the difference between the inner join and the left outer join I'm using. Now, for the `keysummary` aggregation, I don't care about a dataset without metadata; but while `.with_entities` allows me to override the original `SELECT` columns, I'd have to rewrite the base `_get` code to change the `JOIN` type (a default inner join would be fine here). But since the only downside is that we get these "phantom" rows of null key/value for datasets without metadata, and on a real server that's extremely rare if not outright impossible (even for a dataset without a `metadata.log` we'll have `server.archiveonly` and `server.tarball_path`), it wasn't worth worrying about.

I think that realistically this is a unit-test-only case. I did consider adding metadata for all the remaining datasets so I theoretically didn't have to worry about it: but covering the case "to be safe" seemed like a better strategy anyway, and it shouldn't cost much.

Reply:
> covering the case "to be safe" seemed like a better strategy anyway, and it shouldn't cost much.

I concur.  I just wanted to understand how the situation could arise.

Thanks!